app/src/main/java/ch/epfl/sdp/appart/database/local/LocalDatabase.java

```diff
@@ -0,0 +1,880 @@
+            throw new IllegalArgumentException();
+        String favoritesPath = getFavoritesPath();
+
+        //This path should allow multiple users per phone as the userId is
+        // used to name the folder
+        String adFolderPath = getAdFolderPath(cardId);
+
+
+        CompletableFuture<Void> futureSuccess = new CompletableFuture<>();
+        if (!createAdFolder(adFolderPath, ad.getPhotosRefs().size(),
+                ad.getPanoramaReferences().size())) {
+            futureSuccess.completeExceptionally(new LocalDatabaseException(
+                    "Error while creating the ad folder !"));
+            return futureSuccess;
+        }
+
+
+        //I think we do not need to do something specific if the folder
+        // already exists, we will just overwrite the files
+        //Maybe we need to check for the number of images so that if the ad
+        // has one image less than before we do not could keep a stale image
+        // in the folder
+        // in the folder
+
+        //Building local versions of the ad only because we can build the
+        // card from the ad
+        Ad localAd = buildLocalAd(ad, adFolderPath);
+        User localUser = buildLocalUser(user, adFolderPath);
+        Card localCard = buildCardFromAd(localAd, cardId, adId,
+                localUser.getUserId());
+
+        //Adding the data to memory if we have done the first load
+        syncWithMemory(adId, localAd, localCard, localUser);
+
+        //Write the user
+        if (!writeUser(localUser, favoritesPath)) {
+            futureSuccess.completeExceptionally(new LocalDatabaseException(
+                    "Error while writing the user !"));
+            return futureSuccess;
+        }
+
+        //We do not need to check if the set contains it or not.
+        this.userIds.add(localUser.getUserId());
+
+
+        //Serializing
+        if (!writeAd(adId, localAd, localUser.getUserId(), cardId,
+                adFolderPath)) {
+            futureSuccess.completeExceptionally(new LocalDatabaseException(
+                    "Error while writing the ad !"));
+            return futureSuccess;
+        }
+
+        if (!writeAdPhotos(adPhotos, adFolderPath)) {
+            futureSuccess.completeExceptionally(new LocalDatabaseException(
+                    "Error while writing the ad photos !"));
+            return futureSuccess;
+        }
+
+        if (!writePanoramas(panoramas, adFolderPath)) {
+            futureSuccess.completeExceptionally(new LocalDatabaseException(
+                    "Error while writing the ad photos !"));
+            return futureSuccess;
+        }
+        if (loadProfilePic != null && !localUser.hasDefaultProfileImage()) {
+            writeProfilePic(loadProfilePic, localUser.getUserId()).thenAccept(o -> futureSuccess.complete(null)).exceptionally(e -> {
+                futureSuccess.completeExceptionally(e);
+                return null;
```

Comment:
Shouldn't this return `futureSuccess`?

Reply:
This was removed in the latest commit which is a massive refactor.